# Quickstart

In this guide, we will walk through how you can upload a simple Snakemake workflow to Latch.

The example being used here comes from the [short tutorial in Snakemake's documentation](https://snakemake.readthedocs.io/en/stable/tutorial/short.html).

## Prerequisites

- Install a compatible version of Python. The Latch SDK is currently only supported for Python >=3.8 and <=3.11
- Install the [Latch SDK](https://github.com/latchbio/latch#installation) with [snakemake](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html). We recommend installing Latch SDK in a fresh environment for best behaviour:

```console
python3 -m venv env
source env/bin/activate
pip install "latch[snakemake]"
```

## Step 1: Clone the example Snakemake workflow

First, clone the example Snakemake workflow:

```console
git clone <>
```

The workflow generated contains what is typically seen in a Snakemake workflow, such as python scripts and a Snakefile.

```console
snakemake-wf
├── Snakefile
├── environment.yaml
├── config.yaml
├── scripts
│   └── plot-quals.py
```

## Step 2: Add Latch Metadata

The `latch_metadata.py` is used to specify the input parameters and Latch specific metadata that the Snakemake workflow needs to run.

You can automatically generate the `latch_metadata.py` file from an existing `config.yaml` by typing:

```console
latch generate-metadata <path_to_config.yaml>
```

You should find the following files in your workflow directory:

```python
# latch_metadata.py
from latch.types.metadata import SnakemakeMetadata, SnakemakeFileParameter
from latch.types.directory import LatchDir
from latch.types.metadata import LatchAuthor, LatchMetadata, LatchParameter
from pathlib import Path

SnakemakeMetadata(
    display_name="snakemake_tutorial_workflow",
    author=LatchAuthor(
            name="latchbio",
    ),
    parameters={
        "samples" : SnakemakeFileParameter(
                display_name="Sample Input Directory",
                description="A directory full of FastQ files",
                type=LatchDir,
                path=Path("data/samples"),
        ),
        "ref_genome" : SnakemakeFileParameter(
                display_name="Indexed Reference Genome",
                description="A directory with a reference Fasta file and the 6 index files produced from `bwa index`",
                type=LatchDir,
                path=Path("genome"),
        ),
    },
)
```

We will now explain each of the fields in the generated `SnakemakeMetadata` object:

- `display_name`:
- `author`:
- `parameters`:
  For each `SnakemakeFileParameter` parameter, the `path` keyword specifies the path where files will be copied before the Snakemake workflow is run and should match the paths of the inputs for each rule in the Snakefile.

## Step 3: Define Workflow Environment

Next, create an `environment.yaml` file to specify the dependencies that the Snakefile needs to run successfully:

A Dockerfile can be automatically generated by typing:

```console
latch dockerfile snakemake-wf --snakemake
```

## Step 3: Upload the workflow to Latch

Finally, type the following command to register the workflow to Latch:

```console
cd snakemake-wf &&\
latch register . --snakefile Snakefile
```

During registration, a workflow image is built based on dependencies specified in the `environment.yaml` file. Once the registration finishes, the `stdout` provides a link to your workflow on Latch.

![Snakemake workflow interface on Latch](../assets/snakemake/tutorial.png)

## Step 4: Run the workflow

Upload data, and then click execute
Snakemake support is currently uses JIT (Just-In-Time) registration. This means that the workflow produced by `latch register` will register a second workflow, which will run the actual Snakemake jobs.

Once the workflow finishes running, results will be deposited to [Latch Data](https://console.latch.bio/data) under the `Snakemake Outputs` folder.

## Next Steps

- Learn more about the lifecycle of a Snakemake workflow on Latch by reading our [manual](../snakemake/lifecycle.md).
- Learn about how to modify Snakemake workflows to be cloud-compatible [here](../snakemake/cloud.md).
- Visit [troubleshooting](../snakemake/troubleshooting.md) to diagnose and find solutions to common issues.
- Visit the repository of [public examples](https://github.com/latchbio/latch-snakemake-examples) of Snakemake workflows on Latch.
